src/server/get_scale.rs

```diff
@@ -0,0 +1,149 @@
+                    data.xdg_output_manager = Some(xdg_output_manager);
+                },
+                _ => {
+                },
+            }
+        }
+    }
+}
+
+impl Dispatch<wl_output::WlOutput, ()> for AppData {
+    fn event(
+        data: &mut Self,
+        _: &wl_output::WlOutput,
+        event: wl_output::Event,
+        _: &(),
+        _: &Connection,
+        _: &QueueHandle<AppData>,
+    ) {
+        if let wl_output::Event::Mode { flags, width, height, refresh } = event {
+            data.physical_width = width;
+            data.physical_height = height;
+        }
+    }
+}
+
+impl Dispatch<ZxdgOutputManager, ()> for AppData {
+    fn event(
+        _: &mut Self,
+        _: &ZxdgOutputManager,
+        _: ZxdgOutputManagerEvent,
+        _: &(),
+        _: &Connection,
+        _: &QueueHandle<AppData>,
+    ) {
+    }
+}
+
+impl Dispatch<ZxdgOutput, ()> for AppData {
+    fn event(
+        data: &mut Self,
+        _: &ZxdgOutput,
+        event: ZxdgOutputEvent,
+        _: &(),
+        _: &Connection,
+        _: &QueueHandle<AppData>,
+    ) {
+        if let ZxdgOutputEvent::LogicalSize { width, height } = event {
+            data.logical_width = width;
+            data.logical_height = height;
+        }
+    }
+}
+
+// The main function of our program
+pub fn get_scale() -> f64 {
+    let mut scale = 1.0;
+    // Create a Wayland connection by connecting to the server through the
+    // environment-provided configuration.
+    let conn = Connection::connect_to_env();
+    if let Ok(conn) = conn {
+        // Retrieve the WlDisplay Wayland object from the connection. This object is
+        // the starting point of any Wayland program, from which all other objects will
+        // be created.
+        let display = conn.display();
+
+        // Create an event queue for our event processing
+        let mut event_queue = conn.new_event_queue();
+        // And get its handle to associated new objects to it
+        let qh = event_queue.handle();
+
+        // Create a wl_registry object by sending the wl_display.get_registry request
+        // This method takes two arguments: a handle to the queue the newly created
+        // wl_registry will be assigned to, and the user-data that should be associated
+        // with this registry (here it is () as we don't need user-data).
+        let _registry = display.get_registry(&qh, ());
+
+        let mut data = AppData::default();
+
+        let ret = event_queue.blocking_dispatch(&mut data);
+        if ret.is_err() {
+            return scale;
+        }
+
+        if let (Some(wl_output), Some(xdg_output_manager)) = (&data.wl_output, &data.xdg_output_manager) {
+            xdg_output_manager.get_xdg_output(&wl_output, &qh, ());
+        }
+
+        let ret = event_queue.blocking_dispatch(&mut data);
+        if ret.is_err() {
+            return scale;
+        }
+
+        scale = data.physical_width as f64 / data.logical_width as f64;
```

Comment:
what will happen if data.logical_width == 0?

Reply:
Will get NaN according to IEEE.

Reply:
It's avoid.